src/expr/node_traversal.cpp

```diff
@@ -0,0 +1,160 @@
+/*********************                                                        */
+/*! \file node_traversal.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Alex Ozdemir
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2019 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Iterators for traversing nodes.
+ **/
+
+#include "node_traversal.h"
+
+namespace CVC4 {
+
+NodeDfsIterator::NodeDfsIterator(TNode n, bool postorder)
+    : d_stack{n},
+      d_visited(),
+      d_postorder(postorder),
+      d_initialized(false),
+      d_current(TNode())
+{
+}
+
+NodeDfsIterator::NodeDfsIterator(bool postorder)
+    : d_stack(),
+      d_visited(),
+      d_postorder(postorder),
+      d_initialized(true),
+      d_current(TNode())
+{
+}
+
+NodeDfsIterator& NodeDfsIterator::operator++()
+{
+  // If we were just constructed, advance to first visit
+  initializeIfUninitialized();
+
+  // Void our current visit, and advance to the next one
+  d_current = TNode();
```

Comment:
Is this really necessary?

Reply:
Right now it is needed, but there are some other options.

The problem is that the end iterator has a null `d_current`, so a fully-incremented iterator should also have a null `d_current`.

The idea here is that we nullify `d_current`, and `advanceToNextVisit` will populate `d_current` if there is another node to visit. If not it will leave it as the null node.

A few other options:
* have `advanceToNextVisit` write in a null if needed
* do something fancy with `operator==` that I haven't thought of.

I'm certainly open to any of the above options, and others, if you have ideas.

Reply:
Couldn't we just set `d_current` to `TNode()` after the while loop in `advanceToNextVisit`? That would seem quite natural to me (i.e. `advanceToNextVisit` would always update `d_current`).

Reply:
Yeah, I think that's the first option in the list above. I would support it. What do  you think, Mathias?

Reply:
Sounds good!

Reply:
Done in 836575c38